### ISSUE TITLE
Propsal: r.rename

### ISSUE THREAD

**Issue by mlucy** [rethinkdb/rethinkdb#881]
You should be able to rename attributes like:

``` ruby
> r({:a => 1, :b => 2}).rename(:a => :foo, :b => :bar).run
{'foo' => 1, 'bar' => 2}
```

**Comment by darrrk** [comment]
:+1:

**Comment by Xheotris** [comment]
It's such a small feature, but oh-my-goodness, a quick, concise way to change field names throughout a table would be a miracle.

**Comment by danielmewes** [comment]
This should also work for nested fields.

Proposal:
```rb
> r({:a => {:b => 1}, :c => 2}).rename({:a => {:b => :bar}}).run
{:a => {:bar => 1}}
> r({:a => {:b => 1}, :c => 2}).rename({:a => {:b => :bar}, :a => :foo}).run
{:foo => {:bar => 1}}
```
Note that in this proposal, you can not move a field to a different level. E.g. you couldn't move `a.b` to the top level. I think that's probably fine.

It is also slightly ambiguous in the second example `> r({:a => {:b => 1}, :c => 2}).rename({:a => {:b => :bar}, :a => :foo}).run`
whether we refer to the field `a` by its old name `a`, or by its new name `foo` in the rename rule for `a.b`/`foo.b`. I suggest we always assume the old names (like I did here), so we don't need to imply any arbitrary ordering in which the renames will be applied. Basically all renames would be applied atomically in one step.

I also suggest that we error on circular/conflicting renames (e.g. `{:a => :b, :b => :a}` or `{:a => :b, :b => :c}`).

If a target field name already exists, should we just cobble it? Or error? Or have an optarg `conflict: "error"|"cobble"|"merge"`? Maybe with names consistent with `insert` instead: `"error"|"replace"|"update"`?

**Comment by encryptio** [comment]
Most languages (including our `datum_t`s) won't let you define `{:a => one thing, :a => something else}` and get two pairs out of it. Some clobber, some error.

You could do `r.rename({ :a => { :b => :other } }).rename({:a => :top})` as a (not so pretty) workaround.

**Comment by danielmewes** [comment]
Ouch yeah. I didn't think that through. Thanks for pointing that out @encryptio .

We could make `rename` variadic and allow multiple dictionaries to be passed in. Alternatively we could not do that, and require users to chain multiple `rename`s together if they want to rename field names on multiple levels. That would also avoid the ordering ambiguity.

**Comment by Xheotris** [comment]
I can only comment on the JS syntax, but I would make it as close to the Update syntax as possible, since that's basically what you're doing. Something like: 

    { status: "good" } // Before

    r.table( "posts" )
      .rename( { status: goodness } )
      .run()

    { goodness: "good" } // After

Since this isn't a function that one should call casually, it should roll-back and spit an error by default on a conflict along the lines of: 

    {
        message: "There were a few records that had conflicting field names."
        ids: [ "1stConflictID", "2ndConflictID" ... ] 
    }

However, it would probably pay to make that behavior configurable. Perhaps include an options object that includes configuration for overwriting, skipping, and so on.

**Comment by mlucy** [comment]
I think the current proposal where you can rename nested fields is good.

I think a `conflict` optarg is good, and that the default should be `error`.

I think having `update` be a legal option is confusing, because I associate `update` with mutation more than merging.  I also think that being consistent with `insert` is desirable though.  How would people feel about changing `insert` so it read `insert(conflict: 'merge')` instead of `insert(conflict: 'update')`?  (I think `replace` and `error` are fine names.)

I think `.rename` should be non-destructive, so you would write `r.table('posts').update{|row| row.rename(status: 'goodness')}` rather than `r.table('posts').rename(status: 'goodness')` as in the above post.

**Comment by danielmewes** [comment]
`insert(conflict: 'merge')` reads very nicely to me. I like that.
I think we should change that and keep supporting the old `insert(conflict: 'update')` as an alias for backwards compatibility.

I agree that `rename` should just be a transformation, not a write operation.

**Comment by mlucy** [comment]
OK, that sounds good to me too.

**Comment by Xheotris** [comment]
@danielmewes I would maybe hesitate to use 'merge' for an operation that just overwrites old values, since in some other contexts, merge implies at least some sort of timestamp check or other intelligence behind which value to keep. I think in SQL merge can be used to do quite a few types of conflict resolution. I think that something like `insert(conflict: 'overwrite'|'keep'|'error')` might be more explicit? Sorry, I'm certainly not an expert.

**Comment by danielmewes** [comment]
@Xheotris That makes sense, but I think in the case of ReQL `merge` is fitting here since it's consistent with the `merge` ReQL command that we already have. If the conflicting field is an object, it will actually perform a "merge" operation in that it recurses into the object and only overwrites fields that have the same name.

**Comment by danielmewes** [comment]
So the current proposal is:
```js
> r({:a => 1, :b => 2}).rename({:a => :foo, :b => :bar}).run
{'foo' => 1, 'bar' => 2}
> r({:a => {:b => 1}, :c => 2}).rename({:a => {:b => :bar}}).run
{:a => {:bar => 1}}
```

with an optarg `conflict: 'error'|'merge'|'replace'` that defaults to `error`.

Additionally, we will add an option `'merge'` to the `conflict` optarg of `insert` that's going to be an alias for `update` (and we will mark `update` deprecated).

**Comment by mlucy** [comment]
+1 for that proposal.

**Comment by marshall007** [comment]
Sorry for the feature creep, I'm a little late to the party on this one. I would find it really useful if you could specify an array as the "new key" and have it generate keys on the new object for each item. Example:

```js
r.expr({ list_price: 500 }).rename({ list_price: [ 'a', 'b', 'c' ] })
// -> { "a": 500, "b": 500, "c": 500 }
```

Additionally, it might be pretty useful if `r.rename` accepted a function with the signature `(key, value, row)` so you could do things dynamically. I don't have an actual use-case for that one, but it seems like something you'd expect from ReQL.

**Comment by danielmewes** [comment]
Uww I like the idea of allowing a `function(key, value, row) -> newKey` argument.
I'm less sure about the array syntax. That seems to be a bit outside the "typical" use case for `rename`?

We're formally past the ReQL proposal period for 2.3, so I suggest we settle this on the API from https://github.com/rethinkdb/rethinkdb/issues/881#issuecomment-164962569 for now.

We can pick the discussion up again if it doesn't make it into 2.3, or if we want to extend it later.

**Comment by mlucy** [comment]
Agreed, let's settle this with the simpler API, and once that's in we can open a new ReQL discussions for the extensions @marshall007 proposed.

**Comment by marshall007** [comment]
@danielmewes @mlucy yep, sounds good!

**Comment by marshall007** [comment]
@danielmewes: to elaborate on the array thing, while I agree it's probably not a typical use-case, trying to accomplish the same thing with `merge` is fairly cumbersome. I most often find myself wanting to do this when I'm dealing with grouped streams and need to pivot the `reduction` values into keys.

```js
r.expr({ group: 500, reduction: [ 'a', 'b', 'c' ] })
.merge(row => row('reduction').map(r => [ r, row('group') ]).coerceTo('object'))
.omit('group')
```

Given that it doesn't result in any compromises to the settled proposal, it seems like a fairly good fit here. It also gives `rename` similar semantics to `pluck` style selectors, which I think is why (to me) it feels like something that should work.

**Comment by srh** [comment]
`r.alias` is a proposed alternate name in #6331.

**Comment by Extarys** [comment]
I would like to rename a field while keeping it's value. +1 :D

**Comment by phablulo** [comment]
+1
